File: src/etch/bindings/java/util/test/TestStrStrHashMapSerializer.java
/* $Id$
 *
 * Created by wert on Jun 25, 2007.
 *
 * Copyright (c) 2007 Cisco Systems, Inc. All rights reserved.
 */

package etch.bindings.java.util.test;

import static org.junit.Assert.assertEquals;

import org.junit.Test;

import etch.bindings.java.msg.ImportExportHelper;
import etch.bindings.java.msg.StructValue;
import etch.bindings.java.msg.Type;
import etch.bindings.java.support.Class2TypeMap;
import etch.bindings.java.util.StrStrHashMap;
import etch.bindings.java.util.StrStrHashMapSerializer;

/** Tests StrStrHashMapSerializer */
public class TestStrStrHashMapSerializer
{
	/** @throws Exception */
	@Test
	public void test0() throws Exception
	{
		StrStrHashMap map = new StrStrHashMap();
		testMap( map );
	}

	/** @throws Exception */
	@Test
	public void test1() throws Exception
	{
		StrStrHashMap map = new StrStrHashMap();
		map.put( "a", "1" );
		testMap( map );
	}

	/** @throws Exception */
	@Test
	public void test2() throws Exception
	{
		StrStrHashMap map = new StrStrHashMap();
		map.put( "a", "1" );
		map.put( "b", "2" );
		testMap( map );
	}

	/** @throws Exception */
	@Test
	public void test3() throws Exception
	{
		StrStrHashMap map = new StrStrHashMap();
		map.put( "a", "1" );
		map.put( "b", "2" );
		map.put( "c", "3" );
		testMap( map );
	}

	/** @throws Exception */
	@Test
	public void testNullValue() throws Exception
	{
		StrStrHashMap map = new StrStrHashMap();
		map.put( "a", "1" );
		map.put( "b", null );
		map.put( "c", "3" );
		testMap( map );
	}

	/** @throws Exception */
	@Test
	public void testNullKey() throws Exception
	{
		StrStrHashMap map = new StrStrHashMap();
		map.put( "a", "1" );
		map.put( null, "2" );
		map.put( "c", "3" );
		testMap( map );
	}

	/** @throws Exception */
	@Test
	public void testNullKeyValue() throws Exception
	{
		StrStrHashMap map = new StrStrHashMap();
		map.put( "a", "1" );
		map.put( null, null );
		map.put( "c", "3" );
		testMap( map );
	}
	
	private void testMap( StrStrHashMap map )
	{
//		System.out.println( "---------------------------------------------" );
//		System.out.println( "map = "+map );
		
		Type type = new Type( "map" );
		Class2TypeMap class2type = new Class2TypeMap();
		StrStrHashMapSerializer.init( type, class2type );
		ImportExportHelper helper = type.getImportExportHelper();
		
		StructValue sv = helper.exportValue( map );
//		System.out.println( "sv = "+sv );
		
		assertEquals( sv.type(), type );
		
		StrStrHashMap map2 = (StrStrHashMap) helper.importValue( sv );
//		System.out.println( "map2 = "+map2 );
		
		assertEquals( map, map2 );
	}
}

File: src/etch/bindings/csharp/util/test/TestStrStrHashMapSerializer.cs
using System;
using System.Collections.Generic;
using System.Text;
using NUnit.Framework;
using Etch.Support;
using Etch.Msg;

namespace Etch.Util.Test
{
    [TestFixture]
    public class TestStrStrHashMapSerializer
    {
        private readonly DefaultValueFactory vf = new MyValueFactory();

        [Test]
        public void TestImport0()
        {
            StrStrHashMap map = new StrStrHashMap();
            TestMap( map );
        }

        [Test]
        public void TestImport1()
        {
            StrStrHashMap map = new StrStrHashMap();
            map.Add( "a", "1" );
            TestMap( map );
        }

        [Test]
        public void TestImport2()
        {
            StrStrHashMap map = new StrStrHashMap();
            map.Add( "a", "1" );
            map.Add( "b", "2" );
            TestMap( map );
        }

        [Test]
        public void TestImport3()
        {
            StrStrHashMap map = new StrStrHashMap();
            map.Add( "a", "1" );
            map.Add( "b", "2" );
            map.Add( "c", "3" );
            TestMap( map );
        }

        // no tests for null keys as string? isn't allowed

        public void TestMap( StrStrHashMap map )
        {
            StructValue sv = new StructValue( MyValueFactory.blah );
            StrStrHashMapSerializer.ExportValue( vf, map, sv );
            StrStrHashMap map2 = StrStrHashMapSerializer.ImportValue( vf, sv );
            Assert.AreEqual( map, map2 );
        }

        public class MyValueFactory : DefaultValueFactory
        {
            private readonly static TypeMap types = new TypeMap();

            private readonly static FieldMap fields = new FieldMap();

            public static XType blah = types.Get( "blah" );

            static MyValueFactory()
            {
                DefaultValueFactory.Init( types, fields );
                StrIntHashMapSerializer.Init( fields, blah );
            }

            public override XType GetType( int? id )
            {
                return types.Get( id );
            }

            public override XType GetType( string name )
            {
                return types.Get( name );
            }

            public override ICollection<XType> GetTypes()
            {
                return types.Values();
            }

            public override Field GetField( int? id )
            {
                return fields.Get( id );
            }

            public override Field GetField( string name )
            {
                return fields.Get( name );
            }

            public override ICollection<Field> GetFields()
            {
                return fields.Values();
            }
        }

    }
}
